storey/targets.py

```diff
@@ -888,48 +891,100 @@ def _init(self):
-                    value = batch[0].get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < self._number_of_tags - 1:
-                        b.write(",")
-                b.write(")")
-            b.write(" VALUES ")
-            for record in batch:
-                b.write("(")
-                for column_index in range(self._number_of_tags, len(self._columns)):
-                    value = record.get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < len(self._columns) - 1:
-                        b.write(",")
-                b.write(") ")
-            b.write(";")
-            insert_statement = b.getvalue()
-        self._connection.execute(insert_statement)
+                sql.write(f" USING {self._supertable} TAGS {self._get_params_template(self._number_of_tags)}")
+            sql.write(f" VALUES {self._get_params_template(self._number_of_values)};")
+            return sql.getvalue()
+
+    @classmethod
+    def _get_unsupported_error(cls, value) -> UnsupportedTDEngineTypeError:
+        return cls.UnsupportedTDEngineTypeError(f"Unsupported value type {type(value)} in {cls.__name__}")
+
+    @classmethod
+    def _value_to_tag(cls, value):
+        import taosws
+
+        if isinstance(value, bool):
+            return taosws.bool_to_tag(value)
+        if isinstance(value, int):
+            return taosws.int_to_tag(value)
+        if isinstance(value, float):
+            return taosws.float_to_tag(value)
+        if isinstance(value, str):
+            return taosws.varchar_to_tag(value)
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
+        return tags
+
+    @classmethod
+    def _raw_value_to_column_fun(cls, value) -> Callable:
+        import taosws
+
+        if isinstance(value, datetime.datetime):
+            return taosws.millis_timestamps_to_column
+        if isinstance(value, bool):
+            return taosws.bools_to_column
+        if isinstance(value, int):
+            return taosws.ints_to_column
+        if isinstance(value, float):
+            return taosws.floats_to_column
+        if isinstance(value, str):
+            return taosws.nchar_to_column
+        raise cls._get_unsupported_error(value)
+
+    @classmethod
+    def _raw_value_to_value(cls, value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)
+        return value
+
+    def _get_batch_values(self, batch: list[dict]) -> list:
+        values = [[] for _ in range(self._number_of_values)]
+        to_column_funs = []
+        val_names = self._columns[self._number_of_tags :]
```

Comment:
I don't think "value names" is correct nomenclature.
```suggestion
        col_names = self._columns[self._number_of_tags :]
```
or `column_names`. Or `regular_column_names` I suppose.